File: fsharp/bob/Bob.fs
module Bob
open System
let isUpper (s:string): bool = s = s.ToUpper()
let isQuestion (s:string): bool = s.EndsWith("?")
let isNumeric (s:string): bool = s |> Seq.exists Char.IsLetter |> not
let isEmpty (s:string): bool = s = ""
let response (input: string): string =
    match input.Trim() with
        | i when i |> isEmpty -> "Fine. Be that way!"
        | i when i |> isNumeric && i |> isQuestion -> "Sure."
        | i when i |> isNumeric -> "Whatever."
        | i when i |> isUpper && i |> isQuestion -> "Calm down, I know what I'm doing!"
        | i when i |> isQuestion -> "Sure."
        | i when i |> isUpper -> "Whoa, chill out!"
        | _ -> "Whatever."

File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    cond do
      silent?(input) -> "Fine. Be that way!"
      yelling?(input) and question?(input) -> "Calm down, I know what I'm doing!"
      yelling?(input) -> "Whoa, chill out!"
      question?(input) -> "Sure."
      true -> "Whatever."
    end
  end

  defp silent?(input), do: String.trim(input) == ""

  defp question?(input), do: String.ends_with?(String.trim(input), "?")

  defp yelling?(input) do
    letters = Regex.replace(~r/[^\p{L}]/u, input, "")
    letters != "" and letters == String.upcase(letters)
  end
end
